M2/Macaulay2/d/actors5.d

```diff
@@ -13,26 +14,44 @@ getParsing(e:Expr):Expr := (
      else nullE);
 setupfun("getParsing",getParsing);
 
-LongDoubleRightArrowFun(lhs:Code,rhs:Code):Expr := binarymethod(lhs,rhs,LongDoubleRightArrowS);
-setup(LongDoubleRightArrowS,LongDoubleRightArrowFun);
-
-LongLongDoubleRightArrowFun(lhs:Code,rhs:Code):Expr := binarymethod(lhs,rhs,LongLongDoubleRightArrowS);
-setup(LongLongDoubleRightArrowS,LongLongDoubleRightArrowFun);
-
-LongDoubleLeftArrowFun1(rhs:Code):Expr := unarymethod(rhs,LongDoubleLeftArrowS);
-LongDoubleLeftArrowFun2(lhs:Code,rhs:Code):Expr := binarymethod(lhs,rhs,LongDoubleLeftArrowS);
-setup(LongDoubleLeftArrowS,LongDoubleLeftArrowFun1,LongDoubleLeftArrowFun2);
-
-LongLongDoubleLeftArrowFun1(rhs:Code):Expr := unarymethod(rhs,LongLongDoubleLeftArrowS);
-LongLongDoubleLeftArrowFun2(lhs:Code,rhs:Code):Expr := binarymethod(lhs,rhs,LongLongDoubleLeftArrowS);
-setup(LongLongDoubleLeftArrowS,LongLongDoubleLeftArrowFun1,LongLongDoubleLeftArrowFun2);
-
-LongBiDoubleArrowFun(lhs:Code,rhs:Code):Expr := binarymethod(lhs,rhs,LongBiDoubleArrowS);
-setup(LongBiDoubleArrowS,LongBiDoubleArrowFun);
-
-binaryDeductionFun(lhs:Code,rhs:Code):Expr := binarymethod(lhs,rhs,DeductionS);
-unaryDeductionFun(rhs:Code):Expr := unarymethod(rhs,DeductionS);
-setup(DeductionS,unaryDeductionFun,binaryDeductionFun);
+export makeKeywordFun(e:Expr):Expr := (
+    when e
+    is seq:Sequence do
+    if length(seq) != 4
+	  then WrongNumArgs(4)
+	  else (
+     	when seq.0
+     	is s:stringCell do
+	when seq.1
+	is p:ZZcell do
+	when seq.2
+        is b1:Boolean do
+	when seq.3
+	is b2:Boolean do (
+	   if !isvalidkeyword(s.v) then buildErrorPacket("invalid keyword") else (
+	   u:=errorunary;
+	   t:=errorbinary;
+	   prec:=toInt(p);
+	   uprec:=nopr;
+	   bprec:=nopr;
+	   if b1.v || b2.v then (
+	   if b1.v then (t=binaryop; bprec=prec);
+	   if b2.v then (u=unaryop; uprec=prec);
+	   ) else t=postfixop;
+	   w:=makeUniqueWord(s.v, parseinfo(prec,bprec,uprec,parsefuns(u,t)));
+	   when globalLookup(w) is x:Symbol do buildErrorPacket("symbol already in use")
+	   else (
+	   install(s.v,w);  -- TODO check whether install is really needed (for mathematical symbols as opposed to words)
```

Comment:
Are there any updates on whether we need to call `install` here?

Reply:
yeah this is rather subtle. my understanding is, it's harmless to always have `install` here (except maybe for a tiny slowdown of the parsing?). The point is, this is useless for keywords that are actual words (like `and`) because you need to separate them with spaces anyway (so they behave more similarly to regular symbols). But at the moment I've put limited constraints on allowed keywords (see `isvalidkeyword` -- which may require more restrictions), so people can have keywords like `@*@` for which `install` is definitely necessary.

Reply:
I rephrased slightly the comment, but haven't actually changed anything

Reply:
I need to think about this some more. Relatedly, at this stage nothing would prevent us from removing the section in `lex.d` which you wrote `else if ismathoperator(peek2(file)) then (` etc except maybe unicode synonyms.

Reply:
answering myself: no that doesn't work. math operators now have a slightly ambiguous status of being both "alphanumeric" (i.e., usable in a regular symbol) and have special parsing rules (so they can be used in keywords, e.g., binary operators).

Reply:
it would be so much easier if we could declare math symbols to be non alphanumeric. In fact this was my original idea in my first PR on unicode symbols, but that breaks an existing package which defined the sum and product unicode symbols, so I had to change it. now we're paying the price for this backwards compatibility.

Reply:
What package is this?

Reply:
I think it's MultiprojectiveVarieties, which defines `∏`, `⋂`, and `⋃` methods.  I think they'd probably still work as unary operators created using `makeKeyword` after this PR, right?

Reply:
well, right now, this PR doesn't change the status of math operators, so everything works fine as written.
but yes, as you pointed out, a nice feature of this PR is one can even define unary (prefix) operators which for all practical purposes work the same as a `MethodFunction`. so there would definitely be one way forward. might want to have a separate PR for this whole mess. In the meantime, I will fix the `install` thing today so we can resolve this particular comment.

Reply:
> it would be so much easier if we could declare math symbols to be non alphanumeric. In fact this was my original idea in my first PR on unicode symbols, but that breaks an existing package which defined the sum and product unicode symbols, so I had to change it. now we're paying the price for this backwards compatibility.

I don't follow this then. How does declaring math symbols to be non-alphanumeric break backwards compatibility with a package that uses non-alphanumeric method names? Unless you meant declare any non-alphanumeric string to be a math operator automatically?

Reply:
I agree that we should aim for the ability to declare different types of operators, binary, unary left/right, with prescribed precedents.

Reply:
> > it would be so much easier if we could declare math symbols to be non alphanumeric. In fact this was my original idea in my first PR on unicode symbols, but that breaks an existing package which defined the sum and product unicode symbols, so I had to change it. now we're paying the price for this backwards compatibility.
> 
> I don't follow this then. How does declaring math symbols to be non-alphanumeric break backwards compatibility with a package that uses non-alphanumeric method names? 
you mean, with a package that uses math-operator method names?
because methods have ordinary symbols (as opposed to keywords), and these are required to be alphanumeric.